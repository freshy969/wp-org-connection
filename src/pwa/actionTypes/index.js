export const ENTITY_REQUESTED = 'connection/ENTITY_REQUESTED';
export const ENTITY_SUCCEED = 'connection/ENTITY_SUCCEED';
export const ENTITY_FAILED = 'connection/ENTITY_FAILED';

export const LIST_REQUESTED = 'connection/LIST_REQUESTED';
export const LIST_SUCCEED = 'connection/LIST_SUCCEED';
export const LIST_FAILED = 'connection/LIST_FAILED';

export const CUSTOM_REQUESTED = 'connection/CUSTOM_REQUESTED';
export const CUSTOM_SUCCEED = 'connection/CUSTOM_SUCCEED';
export const CUSTOM_FAILED = 'connection/CUSTOM_FAILED';

export const ROUTE_CHANGE_REQUESTED = 'connection/ROUTE_CHANGE_REQUESTED';
export const ROUTE_CHANGE_SUCCEED = 'connection/ROUTE_CHANGE_SUCCEED';
export const ROUTE_CHANGE_FAILED = 'connection/ROUTE_CHANGE_FAILED';
export const PREVIOUS_CONTEXT_REQUESTED = 'connection/PREVIOUS_CONTEXT_REQUESTED';

export const SITE_INFO_REQUESTED = 'connection/SITE_INFO_REQUESTED';
export const SITE_INFO_SUCCEED = 'connection/SITE_INFO_SUCCEED';
export const SITE_INFO_FAILED = 'connection/SITE_INFO_FAILED';

export const HEAD_CONTENT_REQUESTED = 'connection/HEAD_CONTENT_REQUESTED';
export const HEAD_CONTENT_SUCCEED = 'connection/HEAD_CONTENT_SUCCEED';
export const HEAD_CONTENT_FAILED = 'connection/HEAD_CONTENT_FAILED';

export const MOVE_ITEM_TO_COLUMN = 'connection/MOVE_ITEM_TO_COLUMN';
export const REPLACE_CONTEXT = 'connection/REPLACE_CONTEXT';
export const ADD_ITEM_TO_COLUMN = 'connection/ADD_ITEM_TO_COLUMN';
