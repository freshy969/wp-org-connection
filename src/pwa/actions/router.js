import * as types from '../types';

export const routeChangeRequested = ({ currentType, currentId, currentList = 'currentList' }) => ({
  type: types.ROUTE_CHANGE_REQUESTED,
  currentType,
  currentId,
  currentList,
});
export const routeChangeSucceed = ({
  pathname,
  currentType,
  currentId,
  currentList = 'currentList',
}) => ({
  type: types.ROUTE_CHANGE_SUCCEED,
  pathname,
  currentType,
  currentId,
  currentList,
});
