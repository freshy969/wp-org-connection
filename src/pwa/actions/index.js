import * as actionTypes from '../actionTypes';

const parse = id => (Number.isFinite(parseInt(id, 10)) ? parseInt(id, 10) : id);

export const singleRequested = ({ singleType, singleId }) => ({
  type: actionTypes.SINGLE_REQUESTED,
  singleType,
  singleId: parse(singleId),
});
export const singleSucceed = ({ singleType, singleId, entities }) => ({
  type: actionTypes.SINGLE_SUCCEED,
  singleType,
  singleId: parse(singleId),
  entities,
});
export const singleFailed = ({ singleType, singleId, error, endpoint }) => ({
  type: actionTypes.SINGLE_FAILED,
  singleType,
  singleId: parse(singleId),
  error,
  endpoint,
});

export const listRequested = ({ listType, listId, page = 1 }) => {
  if (listType === 'latest' && !listId) listId = 'post';
  return {
    type: actionTypes.LIST_REQUESTED,
    listType,
    listId: parse(listId),
    page: parse(page),
  };
};
export const listSucceed = ({
  listType,
  listId = null,
  page = 1,
  total = { entities: 0, pages: 0 },
  result,
  entities,
}) => {
  if (listType === 'latest' && !listId) listId = 'post';
  return {
    type: actionTypes.LIST_SUCCEED,
    listType,
    listId: parse(listId),
    page: parse(page),
    total,
    result,
    entities,
  };
};
export const listFailed = ({ listType, listId, page = 1, error, endpoint }) => ({
  type: actionTypes.LIST_FAILED,
  listType,
  listId: parse(listId),
  page: parse(page),
  error,
  endpoint,
});

export const customRequested = ({ url = '/', name, singleType, page = 1, params = {} }) => ({
  type: actionTypes.CUSTOM_REQUESTED,
  url,
  name,
  singleType,
  page: parse(page),
  params,
});
export const customSucceed = ({
  url = '/',
  name,
  singleType,
  params = {},
  page = 1,
  total = { entities: 0, pages: 0 },
  result,
  entities,
}) => ({
  type: actionTypes.CUSTOM_SUCCEED,
  url,
  name,
  singleType,
  params,
  page: parse(page),
  total,
  result,
  entities,
});
export const customFailed = ({
  url = '/',
  name,
  singleType,
  params = {},
  page = 1,
  error,
  endpoint,
}) => ({
  type: actionTypes.CUSTOM_FAILED,
  url,
  name,
  singleType,
  params,
  page: parse(page),
  error,
  endpoint,
});

export const routeChangeRequested = ({
  selected: { listType, listId, page, singleType, singleId },
  method = 'push',
  context = null,
}) => {
  if (listType && !page) page = 1;
  if (listType === 'latest' && !listId) listId = 'post';
  const selected = {};
  if (listType) selected.listType = listType;
  if (listId) selected.listId = parse(listId);
  if (page) selected.page = parse(page);
  if (singleType) selected.singleType = singleType;
  if (singleId !== undefined) selected.singleId = parse(singleId);
  return {
    type: actionTypes.ROUTE_CHANGE_REQUESTED,
    selected,
    method,
    context,
  };
};
export const routeChangeSucceed = ({
  selected: { listType, listId, page, singleType, singleId },
  method = 'push',
  context = null,
}) => {
  if (listType && !page) page = 1;
  if (listType === 'latest' && !listId) listId = 'post';
  const selected = {};
  if (listType) selected.listType = listType;
  if (listId) selected.listId = parse(listId);
  if (page) selected.page = parse(page);
  if (singleType) selected.singleType = singleType;
  if (singleId !== undefined) selected.singleId = parse(singleId);
  return {
    type: actionTypes.ROUTE_CHANGE_SUCCEED,
    selected,
    method,
    context,
  };
};
export const routeChangeFailed = () => ({
  type: actionTypes.ROUTE_CHANGE_FAILED,
});

export const siteInfoRequested = () => ({
  type: actionTypes.SITE_INFO_REQUESTED,
});

export const siteInfoSucceed = ({ home: { title, description }, perPage }) => ({
  type: actionTypes.SITE_INFO_SUCCEED,
  home: {
    title,
    description,
  },
  perPage: parseInt(perPage, 10),
});

export const siteInfoFailed = ({ error }) => ({
  type: actionTypes.SITE_INFO_FAILED,
  error,
});
