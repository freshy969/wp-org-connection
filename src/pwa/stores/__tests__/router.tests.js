import Connection from '../';
import * as actionTypes from '../../actionTypes';
import * as actions from '../../actions';

describe('Router', () => {
  let connection;
  beforeEach(() => {
    connection = Connection.create({});
  });

  const toPost = ({ id, type, method, context }) => {
    const payload = { selected: { singleType: type || 'post', singleId: id } };
    if (method) payload.method = method;
    if (context) payload.context = context;
    return actions.routeChangeSucceed(payload);
  };

  const toList = ({ id, type, method, context }) => {
    const payload = { selected: { listType: type || 'category', listId: id } };
    if (method) payload.method = method;
    if (context) payload.context = context;
    return actions.routeChangeSucceed(payload);
  };

  const context1 = {
    infinite: false,
    items: [
      { singleType: 'post', singleId: 60 },
      { singleType: 'post', singleId: 68 },
      { singleType: 'post', singleId: 70 },
      [
        { singleType: 'post', singleId: 90 },
        { singleType: 'post', singleId: 98 },
        { singleType: 'post', singleId: 99 },
      ],
    ],
  };

  const context2 = {
    infinite: false,
    items: [
      [{ singleType: 'post', singleId: 60 }, { listType: 'latest', listId: 'post' }],
      { listType: 'category', listId: 12 },
      { listType: 'category', listId: 13 },
      { listType: 'category', listId: 14 },
      { listType: 'category', listId: 20 },
      { listType: 'category', listId: 22 },
    ],
  };

  const context3 = {
    infinite: false,
    items: [
      [{ singleType: 'post', singleId: 60 }, { listType: 'latest', listId: 'post' }],
      { singleType: 'post', singleId: 68 },
      { singleType: 'post', singleId: 70 },
      { singleType: 'post', singleId: 90 },
      { singleType: 'post', singleId: 98 },
      { singleType: 'post', singleId: 99 },
    ],
  };

  const contextAll = {
    infinite: false,
    items: [
      [{ singleType: 'post', singleId: 60 }, { listType: 'latest', listId: 'post' }],
      { listType: 'category', listId: 12, extract: true },
    ],
  };

  const contextEmpty = {
    infinite: false,
    items: [],
  };

  test('initializes empty', () => {
    expect(connection.contexts).toMatchSnapshot();
    expect(connection.context).toBeNull();
  });

  test('initializes items from context and selected', () => {
    connection[actionTypes.ROUTE_CHANGE_REQUESTED](toPost({ id: 160, context: contextAll }));
    expect(connection.single.post[160]).toMatchSnapshot();
    expect(connection.single.post[60]).toMatchSnapshot();
    expect(connection.single.category[12]).toMatchSnapshot();
    expect(connection.list.category[12].page[0]).toMatchSnapshot();
    expect(connection.list.latest.post.page[0]).toMatchSnapshot();
  });

  test('creates a new context from selected post', () => {
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 60 }));
    expect(connection.selected.type).toBe('post');
    expect(connection.selected.id).toBe(60);
    expect(connection.context.index).toBe(0);
    expect(connection.context.column.selected.id).toBe(60);
    expect(connection.context.columns[0].selected.id).toBe(60);
    expect(connection.contexts[0].columns[0].selected.id).toBe(60);
  });

  test('creates a new context from selected list', () => {
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toList({ id: 7 }));
    expect(connection.selected.type).toBe('category');
    expect(connection.selected.id).toBe(7);
    expect(connection.selected.page).toBe(1);
    expect(connection.context.index).toBe(0);
    expect(connection.context.column.selected.id).toBe(7);
    expect(connection.context.columns[0].selected.id).toBe(7);
    expect(connection.contexts[0].columns[0].selected.id).toBe(7);
  });

  test('creates another context from selected', () => {
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 60 }));
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 70 }));
    // Current item
    expect(connection.context.index).toBe(1);
    expect(connection.selected.id).toBe(70);
    expect(connection.selected.type).toBe('post');
    expect(connection.context.column.selected.id).toBe(70);
    expect(connection.context.columns[0].selected.id).toBe(70);
    expect(connection.contexts[1].columns[0].selected.id).toBe(70);
    // Previous item
    expect(connection.contexts[0].columns[0].selected.id).toBe(60);
  });

  test('creates the context and selects the item as specified', () => {
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 60, context: context1 }));
    expect(connection.selected.id).toBe(60);
    expect(connection.selected.type).toBe('post');
    expect(connection.context.columns[3].selected.id).toBe(90);
    expect(connection.contexts[0].columns[3].items[2].id).toBe(99);
    expect(connection.selected.next.id).toBe(68);
    expect(connection.context.columns[2].selected.next.id).toBe(90);
    expect(connection.context.columns[3].selected.next.id).toBe(98);
  });

  test("changes selected when context exists and has 'selected' inside", () => {
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 60, context: context1 }));
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 90 }));
    expect(connection.selected.id).toBe(90);
    expect(connection.selected.type).toBe('post');
  });

  test("changes selected when contexts are equal and have 'selected' inside", () => {
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 60, context: context1 }));
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 90, context: context1 }));
    expect(connection.selected.id).toBe(90);
    expect(connection.selected.type).toBe('post');
    expect(connection.contexts.length).toBe(1);
  });

  test("Moves selected using replace when context exists and has 'selected' inside", () => {
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 60, context: context1 }));
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 70, method: 'replace' }));
    expect(connection.selected.id).toBe(70);
    expect(connection.selected.type).toBe('post');
    expect(connection.selected.next.id).toBe(68);
    expect(connection.context.column.items[0].next.id).toBe(70);
    expect(connection.context.columns.length).toBe(3);
    expect(connection.context.column).toBe(connection.context.columns[0]);
    expect(connection.context.columns[0].items.length).toBe(2);
  });

  test('Replaces context using replace', () => {
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 60, context: context2 }));
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](
      toList({ id: 'post', type: 'latest', method: 'replace', context: context3 }),
    );
    expect(connection.selected.id).toBe('post');
    expect(connection.selected.type).toBe('latest');
    expect(connection.contexts.length).toBe(1);
    expect(connection.context.column.items.length).toBe(2);
    expect(connection.selected.next.type).toBe('post');
  });

  test('Goes back and forward as expected', () => {
    // Init context and then changes selected in the current context
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 60, context: context1 }));
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 90 }));
    expect(connection.contexts.length).toBe(1);

    // Creates a new context
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 190, context: contextEmpty }));
    expect(connection.contexts.length).toBe(2);
    expect(connection.context.index).toBe(1);
    expect(connection.selected.id).toBe(190);

    // Goes back
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 90, method: 'back' }));
    expect(connection.contexts.length).toBe(2);
    expect(connection.context.index).toBe(0);
    expect(connection.selected.id).toBe(90);

    // Goes back again
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 60, context: context1 }));
    expect(connection.context.index).toBe(0);
    expect(connection.selected.id).toBe(60);

    // Goes forward
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](toPost({ id: 90, method: 'forward' }));
    expect(connection.context.index).toBe(0);
    expect(connection.selected.id).toBe(90);

    // Goes forward again
    connection[actionTypes.ROUTE_CHANGE_SUCCEED](
      toPost({ id: 190, method: 'forward', context: contextEmpty }),
    );
    expect(connection.context.index).toBe(1);
    expect(connection.selected.id).toBe(190);
  });
});
