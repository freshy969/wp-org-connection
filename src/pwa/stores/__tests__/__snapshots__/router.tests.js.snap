// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router initializes empty 1`] = `Array []`;

exports[`Router initializes items from context and selected 1`] = `
Object {
  "_link": null,
  "author": null,
  "content": null,
  "creationDate": null,
  "excerpt": null,
  "featured": null,
  "fetching": false,
  "guid": null,
  "id": 160,
  "link": Object {},
  "meta": Object {
    "canonical": null,
    "description": null,
    "pretty": false,
    "title": null,
  },
  "modificationDate": null,
  "ready": false,
  "slug": null,
  "target": null,
  "taxonomiesMap": Object {},
  "title": null,
  "type": "post",
}
`;

exports[`Router initializes items from context and selected 2`] = `
Object {
  "_link": null,
  "author": null,
  "content": null,
  "creationDate": null,
  "excerpt": null,
  "featured": null,
  "fetching": false,
  "guid": null,
  "id": 60,
  "link": Object {},
  "meta": Object {
    "canonical": null,
    "description": null,
    "pretty": false,
    "title": null,
  },
  "modificationDate": null,
  "ready": false,
  "slug": null,
  "target": null,
  "taxonomiesMap": Object {},
  "title": null,
  "type": "post",
}
`;

exports[`Router initializes items from context and selected 3`] = `
Object {
  "_link": null,
  "id": 12,
  "link": Object {},
  "meta": Object {
    "canonical": null,
    "description": null,
    "pretty": false,
    "title": null,
  },
  "name": null,
  "slug": null,
  "target": null,
  "taxonomy": "category",
}
`;

exports[`Router initializes items from context and selected 4`] = `
Object {
  "entities": Array [],
  "fetching": false,
  "ready": false,
}
`;

exports[`Router initializes items from context and selected 5`] = `
Object {
  "entities": Array [],
  "fetching": false,
  "ready": false,
}
`;
